### DISCUSSION TITLE
How to create a push button that works with spacebar?

### DISCUSSION THREAD

**Discussion by jcsxdev** [gtk-rs/gtk4-rs#1308] [Q&A] [answered]
I created a push button using [GestureClick](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/struct.GestureClick.html):

![screencast1](https://user-images.githubusercontent.com/103332557/219905853-f1517b61-c699-4890-9f1d-c92f6ae6711d.gif)

Here is a snippet of the code:

```rust
    // --snip--

    // Add an action when the button is pressed
    gesture_click.connect_pressed(
        clone! (@weak push_button => move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            push_button.set_label("Release me!");
            print!("Button pressed!  \r");
            let _ = std::io::stdout().flush();
        }),
    );

    // Add an action when the button is released
    gesture_click.connect_released(
        clone! (@weak push_button => move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            push_button.set_label("Press me!");
            print!("Button released!\r");
            let _ = std::io::stdout().flush();
        }),
    );

    // --snip--
```
But I can’t figure out how to get this to work with the space bar. Currently, this is what happens using the spacebar:

![screencast2](https://user-images.githubusercontent.com/103332557/219906239-cded2f0e-dc2c-45d3-b693-5a69820c3e34.gif)

I also tried working with the spacebar using [EventControllerKey](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.EventControllerKey.html), but I ran into the same situation as described here:

- [rust - Detect that user presses spacebar GTK4 - Stack Overflow](https://stackoverflow.com/questions/73846699/detect-that-user-presses-spacebar-gtk4)

**Comment by jf2048** [comment]
It should work if you simply use `connect_clicked` on the button itself.

Generally you do not want to use `GestureClick` or `EventControllerKey` unless you are implementing your own widget.

**Reply by jcsxdev** [reply]
Hi @jf2048,

Thank you for your reply and sorry for the late response.

> It should work if you simply use `connect_clicked` on the button itself.

By simply using `connect_clicked` I can detect that the button was pressed, but I can't detect when it was released. As of now, there is no `connect_released` method in [`ButtonExt`](https://docs.rs/gtk4/latest/gtk4/prelude/trait.ButtonExt.html#provided-methods).

My goal is to keep the button pressed while the mouse button is pressed or while the space bar key is pressed. Can you please provide an example using `button_clicked` to detect these two button states?

In the meantime, I managed to make it work using [`GestureClick`](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/struct.GestureClick.html) and [`EventControllerKey`](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/struct.EventControllerKey.html):

- [Solution using `GestureClick` and `EventControllerKey`](#solution_using_gestureclick_and_eventcontrollerkey)

But the following problem occurs when the space bar key is held down:

![GIF01_2025-01-12](https://github.com/user-attachments/assets/23e832dc-87bd-4571-847a-1ee4b934eb12)


As demonstrated in the GIF, the button opacity is constantly toggled when the space bar key is pressed.

The problem does not occur when the mouse button is pressed:

![GIF02_2025-01-12](https://github.com/user-attachments/assets/e551379a-e7ec-4be2-8c28-2a9f420c54a4)

---

## Solution using `GestureClick` and `EventControllerKey` <a name="solution_using_gestureclick_and_eventcontrollerkey"></a>

Current solution:

```rust
    // --snip--

    // Create a GestureClick to handle mouse press and release events
    let gesture_click = gtk::GestureClick::new();

    // Add an action when the button is pressed
    gesture_click.connect_pressed(clone!(
        #[weak]
        push_button,
        move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            push_button.set_label("Release me!");
            print!("Button pressed!  \r");
            let _ = io::stdout().flush();
        }
    ));

    // Add an action when the button is released
    gesture_click.connect_released(clone!(
        #[weak]
        push_button,
        move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            push_button.set_label("Press me!");
            print!("Button released!\r");
            let _ = io::stdout().flush();
        }
    ));

    push_button.add_controller(gesture_click.upcast::<gtk::EventController>());

    // Create a EventControlKey to handle spacebar press and release events
    let controller_key = EventControllerKey::new();

    controller_key.connect_key_pressed(clone!(
        #[weak_allow_none]
        push_button,
        move |_controller, keyval, _keycode, _state| {
            if keyval == Key::space {
                push_button
                    .expect("Failed to unwrap Button")
                    .set_label("Release me!");
                print!("Button pressed!  \r");
                let _ = io::stdout().flush();
            }
            false.into()
        }
    ));

    controller_key.connect_key_released(clone!(
        #[weak]
        push_button,
        move |_controller, keyval, _keycode, _state| {
            if keyval == Key::space {
                push_button.set_label("Press me!");
                print!("Button released!\r");
                let _ = io::stdout().flush();
            }
        }
    ));

    push_button.add_controller(controller_key.upcast::<gtk::EventControllerKey>());

    // --snip--
```

**Comment by jcsxdev** [comment, marked as answer]
To resolve the visual flickering and handle both mouse and spacebar events properly, the core issue to address is the operating system's key autorepeat. When a key is held down, the OS fires continuous "pressed" events.

The key changes made to fix it:

- Added an `is_pressed` flag (`Rc<Cell<bool>>`) to ignore the continuous "pressed" events the OS sends when you hold down the spacebar.
- Manually toggled `gtk::StateFlags::ACTIVE` on the keyboard events so the button actually looks depressed, just like a mouse click.
- Swapped `.expect()` for a safe weak reference check (`weak.upgrade()`) so the app won't panic if the button is dropped.
- Moved the print and flush logic into a small helper function to stop the text from flickering.

Current solution:

```rust
// --snip--

    // State to track if the button is currently "pressed" to ignore OS key autorepeat
    let is_pressed = Rc::new(Cell::new(false));

    // Handle Mouse Events
    let gesture_click = gtk::GestureClick::new();
    gesture_click.connect_pressed(glib::clone!(
        #[weak] push_button,
        #[strong] is_pressed,
        move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            is_pressed.set(true);
            push_button.set_label("Release me!");
            update_status("Button pressed (mouse)!");
        }
    ));

    gesture_click.connect_released(glib::clone!(
        #[weak] push_button,
        #[strong] is_pressed,
        move |gesture, _n_press, _x, _y| {
            gesture.set_state(gtk::EventSequenceState::Claimed);
            is_pressed.set(false);
            push_button.set_label("Press me!");
            update_status("Button released!");
        }
    ));
    push_button.add_controller(gesture_click);

    // Handle Keyboard Events
    let controller_key = gtk::EventControllerKey::new();

    let push_button_weak = push_button.downgrade();
    let is_pressed_clone = is_pressed.clone();
    controller_key.connect_key_pressed(move |_controller, keyval, _keycode, _state| {
        if keyval != gtk::gdk::Key::space {
            return glib::Propagation::Proceed;
        }

        let Some(push_button) = push_button_weak.upgrade() else {
            return glib::Propagation::Proceed;
        };

        if !is_pressed_clone.get() {
            is_pressed_clone.set(true);
            push_button.set_label("Release me!");
            // Set the ACTIVE state flag to give visual feedback (same as mouse click)
            push_button.set_state_flags(gtk::StateFlags::ACTIVE, false);
            update_status("Button pressed (keyboard)!");
        }
        glib::Propagation::Stop
    });

    let push_button_weak = push_button.downgrade();
    let is_pressed_clone = is_pressed.clone();
    controller_key.connect_key_released(move |_controller, keyval, _keycode, _state| {
        if keyval != gtk::gdk::Key::space {
            return;
        }

        let Some(push_button) = push_button_weak.upgrade() else {
            return;
        };

        is_pressed_clone.set(false);
        push_button.set_label("Press me!");
        // Remove the ACTIVE state flag when the key is released
        push_button.unset_state_flags(gtk::StateFlags::ACTIVE);
        update_status("Button released!");
    });
    push_button.add_controller(controller_key);

    // --snip--
```

Result:

<img width="382" height="178" alt="push_button_keyboard" src="https://github.com/user-attachments/assets/9dc72786-954d-4652-8731-2747127622a5" />

<img width="382" height="178" alt="push_button_mouse" src="https://github.com/user-attachments/assets/dd0e87e1-2343-4bb7-be0f-487c102004ac" />
